### ISSUE TITLE
2.4: Unexpected inference of generic

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#16705]
TypeScript 2.4

``` ts
{
    type Left<L, A> = { value: L };
    type Right<L, A> = { value: A };
    type Either<L, A> = Left<L, A> | Right<L, A>;
    const eitherLeft = <L, A>(l: L): Either<L, A> => ({ value: l })
    const eitherRight = <L, A>(a: A): Either<L, A> => ({ value: a })

    const fn1 = (): Either<string, number> => eitherRight(1); // no error
    const fn2 = (): Promise<Either<string, number>> => Promise.resolve(eitherRight(1)); // error
}
```

The full error:

```
Type 'Promise<Either<string | number, number>>' is not assignable to type 'Promise<Either<string, number>>'.
  Type 'Either<string | number, number>' is not assignable to type 'Either<string, number>'.
    Type 'Left<string | number, number>' is not assignable to type 'Either<string, number>'.
      Type 'Left<string | number, number>' is not assignable to type 'Right<string, number>'.
        Types of property 'value' are incompatible.
          Type 'string | number' is not assignable to type 'number'.
            Type 'string' is not assignable to type 'number'.
```

I would not expect this code to error.

`fn2`’s `eitherRight` call is inferred to have generics `<string | number, number>`:

- The first generic infers `string` from the return type of `fn2`
- The second generic infers `number` from its argument

What I don't understand is where the first generic infers `number` from?

**Comment by aleksey-bykov** [comment]
make it
type Left<L> = { left: L; }
type Right<R> = { right: R; }

**Comment by aleksey-bykov** [comment]
https://github.com/aleksey-bykov/basic/blob/master/either.ts

**Comment by OliverJAsh** [comment]
Thanks, but I still don't understand why inference is working this way. It feels like a bug to me.

**Comment by aleksey-bykov** [comment]
1. you started screwing things up by using the same property `value` both for left and right cases, if the properties were named differently you would not see it (like in my example), the reason for this problem is that type script has structural typing it means that out of 2 different declaration of `value` (in Left and in Right) it needs to make 1 that would fit both (in Either) which is `value: L | R` 

2. scondly in the latest typescript unresolved type parameters in the right side are assisted from the declarations on the left side, i am not an expert in how it works in details, but it looks like because there is an unresolved type parameter on the right side (`L` which is `{}`) the inference mechanism is looking for the best candidate for it on the left side (`string` as declared), but because of what's in point 1 above it is also `number` hence you get `string | number` which typescript use to satisfy the left side expectation 

i might as well be wrong about it, long story short - mind structural typing

**Comment by OliverJAsh** [comment]
@aleksey-bykov Thanks for that explanation.

It's still odd that it works if you remove the `Promise`, e.g.

``` ts
    const fn1 = (): Either<string, number> => eitherRight(1); // no error
```

**Comment by OliverJAsh** [comment]
And it seems to infer the generics correctly:

![image](https://user-images.githubusercontent.com/921609/27518766-526e926e-59b5-11e7-9385-71925c83f5b9.png)

But no when you wrap it in `Promise`:

``` ts
    const fn2 = (): Promise<Either<string, number>> => Promise.resolve(eitherRight(1)); // error
```

![image](https://user-images.githubusercontent.com/921609/27518768-5f48bb40-59b5-11e7-9543-434e4dbc7452.png)

**Comment by aleksey-bykov** [comment]
i see what you mean, have no idea, although it's clear that the left side has something to do with it because as you see:
```
const fn1 = () => eitherRight(1); //  is () => { value: {}; } | { value: number; }
const fn2 = () => Promise.resolve(eitherRight(1)); // is () => Promise<{ value: {}; } | { value: number;  }>
```

**Comment by aleksey-bykov** [comment]
there were some special typing rules specifically targeting Promise and interfaces alike, might be a consequences of this: https://github.com/Microsoft/TypeScript/pull/15104

**Comment by ahejlsberg** [comment]
You are running into the limits of what type inference can figure out. In the call `Promise.resolve(eitherRight(1))`, the compiler ends up inferring from the type `Either<string, number> | Promise<Either<string, number>>` (the contextual type for the `eitherRight(1)` argument) to the type `Either<L, A>` (the return type of `eitherRight`).  However, when the union type `Either` is put into another union type, we "flatten" the resulting type to `Left<string, number> | Right<string, number> | Promise<Either<string, number>>`. That in turn causes us to no longer "see" that we're inferring from an `Either` to an `Either`, and so we infer from *each* constituent in the source to *each* constituent in the target which ends up confusing the `L` and `A` inferences. Ideally, inference involving union types should pick a "best match" constituent and then ignore inferences from the others, but it's not clear what the algorithm for doing that would be.

In other words: It's complicated.

**Comment by OliverJAsh** [comment]
@ahejlsberg Appreciate the insight, because it's good to know when I'm working with a constraint of the system, and why that is. :-)

**Comment by OliverJAsh** [comment]
Is it possible to throw an error/warning when inference may be falling short, like it is here? I'm happy to provide type annotations, but I don't want bad inference to cause silent errors in my code.

**Comment by sylvanaar** [comment]
Can't reproduce this in TypeScript 2.4.1

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
